File: backend/app/schemas/outreach.py
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict

from app.models.outreach_template import OutreachCategory, OutreachChannel


class OutreachTemplateBase(BaseModel):
    channel: OutreachChannel
    category: OutreachCategory
    title: str
    subject: str | None = None
    body: str
    notes: str | None = None
    sort_order: int = 0
    is_favorite: bool = False
    # Palette-Schlüssel, nicht Farbwert (s. Modell). None = neutral.
    color: str | None = None
    # Reihenfolge innerhalb der Favoriten (kanalübergreifend, s. Modell).
    favorite_order: int | None = None


class OutreachTemplateCreate(OutreachTemplateBase):
    pass


class OutreachTemplateUpdate(BaseModel):
    channel: OutreachChannel | None = None
    category: OutreachCategory | None = None
    title: str | None = None
    subject: str | None = None
    body: str | None = None
    notes: str | None = None
    sort_order: int | None = None
    is_favorite: bool | None = None
    color: str | None = None
    favorite_order: int | None = None


class OutreachTemplateResponse(OutreachTemplateBase):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    usage_count: int
    created_at: datetime
    updated_at: datetime


class OutreachReorder(BaseModel):
    """Neue Reihenfolge eines Kanals als vollständige ID-Liste.

    Bewusst die ganze Liste und nicht „verschiebe X hinter Y": Der Client kennt
    die Zielordnung ohnehin, und ein Bulk-Setzen ist atomar. Einzelne PUTs pro
    Karte wären N Requests und könnten zur Hälfte scheitern.
    """

    channel: OutreachChannel
    ids: list[uuid.UUID]


class OutreachFavoriteReorder(BaseModel):
    """Neue Reihenfolge der Favoriten als vollständige ID-Liste.

    Ohne `channel`: Die Favoriten-Sektion ist kanalübergreifend. Sie schreibt
    `favorite_order`, nicht `sort_order` — sonst würde ein Zug in den Favoriten
    die Reihenfolge der Kanal-Listen mitverändern.
    """

    ids: list[uuid.UUID]

File: frontend/src/api/outreach.ts
import { api } from './client'
import type {
  OutreachChannel,
  OutreachTemplate,
  OutreachTemplateCreate,
  OutreachTemplateUpdate,
} from '../types'

export async function getOutreachTemplates(): Promise<OutreachTemplate[]> {
  const res = await api.get('/outreach/templates')
  return res.data
}

export async function createOutreachTemplate(
  data: OutreachTemplateCreate,
): Promise<OutreachTemplate> {
  const res = await api.post('/outreach/templates', data)
  return res.data
}

export async function updateOutreachTemplate(
  id: string,
  data: OutreachTemplateUpdate,
): Promise<OutreachTemplate> {
  const res = await api.put(`/outreach/templates/${id}`, data)
  return res.data
}

export async function deleteOutreachTemplate(id: string): Promise<void> {
  await api.delete(`/outreach/templates/${id}`)
}

/** Zählt eine Nutzung (Copy) hoch. */
export async function markOutreachCopied(id: string): Promise<OutreachTemplate> {
  const res = await api.post(`/outreach/templates/${id}/copied`)
  return res.data
}

/** Legt die Standard-Templates an (idempotent) und gibt die aktuelle Liste zurück. */
export async function seedOutreachTemplates(): Promise<OutreachTemplate[]> {
  const res = await api.post('/outreach/templates/seed')
  return res.data
}

/**
 * Setzt die Reihenfolge eines Kanals. Bewusst die vollständige ID-Liste: ein
 * Bulk-Setzen ist atomar, während einzelne PUTs zur Hälfte scheitern und die
 * Liste halb sortiert hinterlassen könnten.
 */
export const reorderOutreachTemplates = async (
  channel: OutreachChannel, ids: string[],
): Promise<OutreachTemplate[]> =>
  (await api.post<OutreachTemplate[]>('/outreach/templates/reorder', { channel, ids })).data

/**
 * Reihenfolge innerhalb der Favoriten. Ohne Kanal — die Sektion ist
 * kanalübergreifend und schreibt `favorite_order`, nicht `sort_order`.
 */
export const reorderOutreachFavorites = async (
  ids: string[],
): Promise<OutreachTemplate[]> =>
  (await api.post<OutreachTemplate[]>('/outreach/templates/favorites/reorder', { ids })).data
